projects/app_update_out_of_date.kp/knowledge.md

```diff
@@ -0,0 +1,1067 @@
+JSON that populates the web page data. This way the reference in the
+web page never changes. A reference is all that is needed since the
+dictionary is not modified.
+
+
+```python
+ofConcernDict = hasUpdateEnabledDict
+```
+Create an RDD of the telemetry pings that have the
+application.update.enabled preference set to True.
+
+This RDD is created from the last "out of date, potentially of concern"
+RDD and it is named ofConcernTrueRDD to simplify the addition of new code
+without having to modify consumers of the RDD.
+
+
+```python
+ofConcernTrueRDD = hasUpdateEnabledRDD.filter(lambda p: p[0] == True).map(lambda p: p[1]).cache()
+```
+Create an RDD of out of date, of concern telemetry ping client
+versions along with a dictionary of the count of each version.
+
+
+```python
+def byVersionMapper(d):
+    ping = d
+    return (ping.version[0], ping)
+
+ofConcernByVersionRDD = ofConcernTrueRDD.map(byVersionMapper)
+ofConcernByVersionDict = ofConcernByVersionRDD.countByKey()
+ofConcernByVersionDict
+```
+Create an RDD of out of date, of concern telemetry ping update check
+codes along with a dictionary of the count of each update check code.
+
+
+```python
+def checkCodeNotifyMapper(d):
+    ping = d
+    index = 0
+    currentVersion = ping.version[0]
+    while (index < len(ping.update_ping_count_notify) and
+           index < len(ping.version) and
+           ping.version[index] == currentVersion):
+        if ping.update_ping_count_notify[index] > 0:
+            if ping.update_check_code_notify is not None:
+                codeIndex = 0
+                for codeValue in ping.update_check_code_notify[index]:
+                    if codeValue > 0:
+                        return (codeIndex, ping)
+                    codeIndex += 1
+
+            if (ping.update_check_no_update_notify is not None and
+                ping.update_check_no_update_notify[index] > 0):
+                return (0, ping)
+
+        index += 1
+
+    return (-1, ping)
+
+checkCodeNotifyOfConcernRDD = ofConcernTrueRDD.map(checkCodeNotifyMapper)
+checkCodeNotifyOfConcernDict = checkCodeNotifyOfConcernRDD.countByKey()
+checkCodeNotifyOfConcernDict
+```
+Create an RDD of out of date, of concern telemetry pings that had a
+general failure for the update check. The general failure codes are:
+* CHK_GENERAL_ERROR_PROMPT: 22
+* CHK_GENERAL_ERROR_SILENT: 23
+
+
+```python
+checkCodeNotifyGeneralErrorOfConcernRDD = \
+  checkCodeNotifyOfConcernRDD.filter(lambda p: p[0] == 22 or p[0] == 23).map(lambda p: p[1]).cache()
+```
+Create an RDD of out of date, of concern telemetry ping update check
+extended error values for the clients that had a general failure for
+the update check along with a dictionary of the count of the error
+values.
+
+
+```python
+def checkExErrorNotifyMapper(d):
+    ping = d
+    currentVersion = ping.version[0]
+    for index, version in enumerate(ping.version):
+        if ping.update_ping_count_notify[index] > 0:
```

Comment:
Fixed the first one but the second one is intentional otherwise it would return even when ping.update_check_extended_error_notify[key_name][index] > 0: was not greater than 0.

if ping.update_check_extended_error_notify[key_name][index] > 0:
    if version == current_version:
        key_name = key_name[17:]
        if len(key_name) == 4:
            key_name = key_name[1:]
    return (int(key_name), ping)
                    return (-1, ping)